mesa/experimental/jupyter_viz.py

```diff
@@ -35,34 +36,66 @@ def JupyterViz(
             simulations with no space to visualize should
             specify `space_drawer=False`
         play_interval: play interval (default: 400)
+        timeline: whether to display a scrubbable timeline (default: False)
     """
 
-    current_step, set_current_step = solara.use_state(0)
-
     # 1. Set up model parameters
     user_params, fixed_params = split_model_params(model_params)
     model_parameters, set_model_parameters = solara.use_state(
         {**fixed_params, **{k: v["value"] for k, v in user_params.items()}}
     )
 
+    model, set_model = solara.use_state(model_class(**model_parameters))
+    model_cache, set_model_cache = solara.use_state({0: model})
+
     # 2. Set up Model
     def make_model():
         model = model_class(**model_parameters)
-        set_current_step(0)
-        return model
+        set_model(model)
+        set_model_cache({0: model})
 
-    reset_counter = solara.use_reactive(0)
-    model = solara.use_memo(
-        make_model, dependencies=[*list(model_parameters.values()), reset_counter.value]
+    solara.use_memo(
+        make_model,
+        dependencies=list(model_parameters.values()),
     )
 
     def handle_change_model_params(name: str, value: any):
         set_model_parameters({**model_parameters, name: value})
 
+    def do_step():
+        if not model.running:
+            return
+
+        updated_model = copy.deepcopy(model)
```

Comment:
You are probably going to say it's fine for small models. But this recreates the grid, agent vars, and model vars from scratch every time. It will hit the resource ceiling sooner than before. Python objects are not meant to be performant for immutable paradigm, are they?

Reply:
Compare with e.g. Agents.jl, which is frugal with resource.

Reply:
> You are probably going to say it's fine for small models. But this recreates the grid, agent vars, and model vars from scratch every time. It will hit the resource ceiling sooner than before. Python objects are not meant to be performant for immutable paradigm, are they?

Yes this is exactly what I am going to say 😅 

Plus, also for large models that should be okay for a long time. For performance there is no right or wrong, but in terms of speed copying is much faster than actually recreating and for memory usage when we deactivate the timeline we only ever hold 2 models at the same time. 

I would say we should try it this way, because it makes the code much simpler and easier to reason about. If we eventually do hit a performance ceiling we can change things. But right now we should optimize for the 80% use case. 

And our solara frontend is still experimental. We should focus on moving forward instead of worrying too much about hypothetical scenarios.

Reply:
Can you describe what features you want to implement with this immutable model convenience? IIRC there was a GH issue where the OP requested for a feature to allow them to change model params (only if it makes sense to do so) without triggering a whole system reset. Is this one of them?

Reply:
Note that some Solara servers are meant to be accessed by lots of users in parallel, and some are run on a small AWS instance. They could be considered to be within 80% of the use case. I have to weigh over this use case vs the features you want to implement.

Reply:
> Can you describe what features you want to implement with this immutable model convenience? IIRC there was a GH issue where the OP requested for a feature to allow them to change model params (only if it makes sense to do so) without triggering a whole system reset. Is this one of them?

 - First of all right now instead of a single "model" variable, we have "model", "current_step" and "reset_counter". Both of the latter are used for triggering change detection, because changes to the model object will go unnoticed as far as solara is concerned. If we implement further model-changing functionalities (ex. interactivity) we always have to worry about keeping model up to date (by adding more and more dependency variables to the make_model function).

- This very PR that adds time traveling to mesa.

- PR #1791 will work without any dummy parameter to function calling trickery.

- Running multiple models in parallel will get messy when we have to store multiple models, model parameters, current_steps and reset_counters. 

- Saving visualization and model state to continue model runs accross sessions. 

These are the features I have currently planned or panned out (mostly following https://github.com/Corvince/mesa-viz). Please don't start any lenghty discussions about any of them.

Reply:
> First of all right now instead of a single "model" variable, we have "model", "current_step" and "reset_counter".

They all can be generalized to at most 2 reactive variables: to inform a model reset (`reset_counter`), and to inform a model update.

```python
model_update = solara.reactive(0)
```
This reactive variable is updated whenever you want to trigger components rerender (e.g. after a `model.step()`).

While `reset_counter` can be removed once `ipywidgets.Play` is replaced.

Reply:
Sorry I have to ask again for clarification. I am really not doing this annoy you. 

What you write is absolutely correct. I just don't know its purpose. Is there a hidden question or some form of conclusion I am not getting or is it a simple statement?

Reply:
I forgot to quote the remaining paragraph

> Both of the latter are used for triggering change detection, because changes to the model object will go unnoticed as far as solara is concerned. If we implement further model-changing functionalities (ex. interactivity) we always have to worry about keeping model up to date (by adding more and more dependency variables to the make_model function).

What I am saying is that the `model_update` variable can be used as a proxy for any interactive features to be based on, if they want to get a redraw signal from model.

Reply:
I understand that, but I am asking what you want to say. Are you seriously opposed to merging this PR, because it copies the model for state updates? This approach incidentally makes the code more straightforward, comprehensible, maintainable and easier to reason about.

Reply:
The current Solara code is already far simpler than mesa-viz-tornado, for what it can already do. Either:
- use `model_update` as a compromise
- we consider your current immutable model implementation as a training wheel so that it is easier for you to experiment with features, but eventually, this must be optimized in the future

Which one do you prefer? It's rather paradoxical that Agents.jl is more performant yet they are more resource conscious, while you are being more extravagant.